lsp.el

```diff
@@ -0,0 +1,511 @@
+;;; lsp.el --- LSP mode                              -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2018  Ivan Yonchovski
+
+;; Author: Ivan Yonchovski <ivan.yonchovski@tick42.com>
+;; Keywords: languages
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;;
+
+;;; Code:
+
+(require 'dash)
+(require 'seq)
+(require 'lsp-mode)
+(require 'f)
+(require 'files)
+(require 'spinner)
+
+(defcustom lsp-auto-guess-root nil
+  "Automatically guess the project root using projectile/project."
+  :group 'lsp
+  :type 'bool)
+
+(defcustom lsp-restart 'interactive
+  "Defines how server exited event must be handled."
+  :group 'lsp
+  :type '(choice (const interactive)
+                 (const auto-restart)
+                 (const ignore)))
+
+(defcustom lsp-session-file (expand-file-name (locate-user-emacs-file ".lsp-session"))
+  "Automatically guess the project root using projectile/project."
+  :group 'lsp
+  :type 'file)
+
+(defcustom lsp-auto-configure t
+  "Auto configure `lsp-mode'.
+When set to t `lsp-mode' will auto-configure `lsp-ui' and `company-lsp'."
+  :group 'lsp
+  :type 'bool)
+
+(defvar lsp-clients (make-hash-table :test 'eql))
+
+(defvar-local lsp--buffer-workspaces ()
+  "List of the buffer workspaces.")
+
+
+(defun lsp-stdio-connection (name command-fn)
+  "Create LSP STDIO connection named NAME.
+COMMAND-FN will be called to generate Language Server command."
+  (lsp--make-stdio-connection name nil command-fn))
+
+(defun lsp--auto-configure ()
+  "Autoconfigure `lsp-ui', `company-lsp' if they are installed."
+  (when (featurep 'lsp-ui)
+    (flycheck-mode)
+    (lsp-ui-mode))
+
+  (when (featurep 'company-lsp)
+    (company-mode-on)
+    (setq company-backends '(comapny-lsp))))
```

Comment:
Typo.

Though personally I prefer a single entry `company-lsp`, some may prefer others.

From my doom-emacs:

```elisp
(company-lsp
 (:separate company-capf company-yasnippet)
 company-capf)
```

Reply:
I am fine with both approaches. This will act like default configuration and if you want something custom you can always turn off lsp-auto-configure and do the configuration by yourself.

Reply:
SG. I don't change the doom-emacs default just because I haven't found other backends get in the way. If they do I'll ruthlessly delete them :)

I think `company-backends` is a defcustom and not buffer-local by default.